ci: add CI for monorepo wp-desktop

nsakaimbo opened the pull request:
### Description

This PR is a follow-on to #43737, adding baseline CI build and e2e for wp-desktop.

- [x] Add upstream source build (`wp-desktop-source`)
- [x] Add Mac build job (`wp-desktop-mac`)
- [x] Add code signing for Mac
- [x] Add e2e tests for Mac
- [x] Local dev environment: add Makefile with convenience utilities and Docker setup (`desktop/Makefile`)

TODO (in downstream PRs separate from this initial pass):
- Mac notarization
- Windows and Linux builds, e2e
- Slack and GH notifications
- application deployment

nsakaimbo committed:
add Makefile and CI source, mac jobs

nsakaimbo committed:
tests: add e2e tests

nsakaimbo committed:
try hack: install desktop dependencies only

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure (ci/wp-desktop): @nsakaimbo please re-try this workflow ("Rerun Workflow from Failed") and/or review this PR for breaking changes. Please also ensure this branch is rebased off latest Calypso master.

nsakaimbo committed:
Try w/out --frozen-lockfile

nsakaimbo committed:
use subprocess to cd and install deps

nsakaimbo commented:
I was curious to inspect the runtime error in wp-desktop. App fails with the message:

```
... cannot find module 'clipboard'
// ...
```

<p align="center">
<img width="302" alt="Screen Shot 2020-06-26 at 12 47 01 PM" src="https://user-images.githubusercontent.com/8979548/85883672-5ddac680-b7af-11ea-8b2b-1bce9f4ab7e6.png">
</p>

nsakaimbo committed:
Revert "use subprocess to cd and install deps"

This reverts commit 850cb4bbf30c45ecfe79cc605b81ad6315357f36.

nsakaimbo committed:
Revert "Try w/out --frozen-lockfile"

This reverts commit 3ee2d04ec83d8cf791b074dfeb5ff5dea226ecf0.

nsakaimbo committed:
Revert "try hack: install desktop dependencies only"

This reverts commit d1f1940b37e18f6ac26eef8eaa820d1ea60b162b.

nsakaimbo committed:
add desktop testing deps to desktop/package.json

nsakaimbo committed:
Invoke e2e tests

nsakaimbo committed:
explicitly set nvm + yarn version before test

nsakaimbo committed:
don't use --ignore-scripts

jsnajdr commented:
Now when #43737 got merged, this branch can be rebased onto the latest _master_.

nsakaimbo committed:
fix desktop e2e dep references

nsakaimbo committed:
fix: use npx for electron-builder

nsakaimbo committed:
add build-desktop:source script

nsakaimbo committed:
use yarn in circle config for consistency

nsakaimbo committed:
ci: install desktop app deps

nsakaimbo committed:
don't use check-npm-client for build-desktop:app

nsakaimbo committed:
remove check-npm-client from install-app-deps

nsakaimbo committed:
restore all jobs

nsakaimbo committed:
add mocha-steps and electron-mocha

nsakaimbo committed:
Revert "restore all jobs"

This reverts commit da7acdc716447ef6f8535cdb93692a0189005ae1.

nsakaimbo committed:
Use max_old_space_size override when building source

nsakaimbo committed:
rename circle jobs

nsakaimbo committed:
restore all CI jobs

nsakaimbo commented on .circleci/config.yml:
```diff
@@ -14,6 +14,15 @@ references:
       PUPPETEER_SKIP_CHROMIUM_DOWNLOAD: 'true'
```
Update: Circle config

nsakaimbo commented on desktop/e2e/run.js:
```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env node
```
Test runner script and test files are unchanged from source in `wp-desktop` repo, save for updating `PROJECT_DIR` and `E2E_DIR` to point to the update folder location.

nsakaimbo commented:
Looks like `ci/wp-e2e-tests-full-desktop` and `ci/wp-e2e-tests-full-mobile` are failing in other branches at the moment as well. Doesn't appear related to the changes in this branch:

```
18:57:31 FAILED: Taking screenshot of: 'https://e2eflow****ing3.wordpress.com/2020/06/29/classy-pumas-clap-rudely/'
18:57:31 Screencast Video:/home/circleci/wp-calypso/wp-calypso/****/e2e/screenshots/videos/can-see-post-message-2020-06-29T18-57-31.mpg
18:57:31   28 passing (undefinedms)
18:57:31   1 failing
18:57:31   1) [WPCOM] Editor: Posts (desktop)
18:57:31        Public Posts: Preview and Publish a Public Post @parallel @jetpack
// ...
```

nsakaimbo commented:
PR is ready for review.

_Highlights_:

- Added Makefile utilities and a Dockerfile (for building source on Windows)
- Lifted existing e2e scripts from wp-desktop repo
- Updated Calypso Circle config to incorporate `wp-desktop-source` and `wp-desktop-mac` jobs (more platforms to follow). Platform build job caches built artifacts regardless of success/failure, making them available for easy inspection.

nsakaimbo committed:
add e2e logs and screenshots to artifacts

nsakaimbo committed:
build-mac: remove unnecessary env var

nsakaimbo committed:
use separate worfklow for wp-desktop

jsnajdr commented on .circleci/config.yml:
```diff
@@ -179,6 +192,40 @@ references:
+    - public
+    - client/server
+    - config/secrets.json
+    - client/desktop/config.json
```
The purpose of this cache is to save the results of the source build and then restore them for the platform-specific packaging jobs?

For that, you should need only the `desktop/` directory contents. `build/` and `public/` shouldn't be created by any `build-desktop:*` task at all, and all the files in `config/` or `client/` should be either copied to `desktop/` by the `build-desktop:config` task, or generated directly into `desktop/`.

nsakaimbo replied:
> The purpose of this cache is to save the results of the source build and then restore them for the platform-specific packaging jobs?

Yes, this is correct. This was tricky to get right, so I ended up caching all places that generated a diff after the node module installation step. Will try limiting the cache paths to just those within the `desktop` folder. 👍

nsakaimbo replied:
Caching only items in `desktop` works - updated in 8f6dca6.

jsnajdr replied:
Seems there is first class CircleCI support for sharing files between jobs: the `persist_to_workspace` and `attach_workspace` commands: https://circleci.com/docs/2.0/workflows/#using-workspaces-to-share-data-among-jobs Can we possibly use these instead of the cache? It can be a followup PR. All my knowledge about this comes from a minute of googling, so I apologize in case I'm suggesting something silly 🙂

nsakaimbo commented on .circleci/config.yml:
```diff
@@ -122,7 +131,11 @@ references:
+    command: |
+      source "$HOME/.nvm/nvm.sh"
+      nvm use
+      npm install -g yarn
```
> It's strange that we need to install yarn like this. Before this PR, where did the yarn executable come from in CircleCI when running yarn-install?

In a nutshell: for each command, Circle essentially uses a brand new shell instance that doesn't necessarily carry over the defaults from a prior instance. You can prove this out by using the `update-node` task and immediately validating what version of node is being used in a subsequent command. Here's where I confirmed this in CI:

_Test: Verify Node Version Without NVM Refresh_: [link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/65090/workflows/9e6046fb-b424-4187-9b52-a5963213bf07/jobs/787705)

```
// command: update-node

#!/bin/bash --login
set +e
set +x
export NVM_DIR="$HOME/.nvm" && (
  git clone https://github.com/nvm-sh/nvm.git "$NVM_DIR"
  cd "$NVM_DIR"
  git checkout v0.35.3
) && \. "$NVM_DIR/nvm.sh" --no-use
nvm install
nvm alias default "$(cat .nvmrc)"
nvm use

// command: Verify Node Version

#!/bin/bash --login
nvm list
yarn --version
/bin/bash: nvm: command not found 
1.12.1

// command: Install dependencies
yarn install v1.12.1
 
error wp-calypso-monorepo@0.17.0: The engine "node" is incompatible with this module. Expected version "^v12.18.0". Got "11.0.0"
error wp-calypso-monorepo@0.17.0: The engine "yarn" is incompatible with this module. Expected version "^1.22.4". Got "1.12.1"
```

You can see that by default, the shell is using Node v11.0.0 and yarn 1.12.1.

_Test: Verify Node Version After NVM Refresh_: [link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/65106/workflows/f584e8f0-dfd8-426f-9c2b-6f0329f2ef69/jobs/787804)

```
// Command: install dependencies

#!/bin/bash --login 
source "$HOME/.nvm/nvm.sh" 
yarn install --frozen-lockfile 
 
yarn install v1.12.1
error wp-calypso-monorepo@0.17.0: The engine "yarn" is incompatible with this module. Expected version "^1.22.4". Got "1.12.1"
error Found incompatible module
```

Now Node version is correct, but yarn is still wrong.

This could be a bug specific to Circle's Mac installation of nvm, but the implication is that nvm has to be refreshed as necessary after the initial setup so all bases are covered. Interestingly enough, this is how it has been wired up in wp-desktop's existing CI as well (see [here](https://github.com/Automattic/wp-desktop/blob/develop/.circleci/config.yml#L409) and [here](https://github.com/Automattic/wp-desktop/blob/develop/.circleci/config.yml#L421)), so this issue must have been encountered before.

nsakaimbo commented on .circleci/config.yml:
```diff
@@ -122,7 +131,11 @@ references:
-    command: yarn install --frozen-lockfile
+    command: |
+      source "$HOME/.nvm/nvm.sh"
+      nvm use
```
See my notes below regarding the default Node version used between Circle commands. `update-node` installs the node version, but at least on Mac it isn't guaranteed that that node version will be used in subsequent commands. `nvm set default <VERSION>` also doesn't seem to work, for some strange reason.

The `yarn-install` command installs dependencies assuming that Node and yarn have already been installed. On Mac, I found that refreshing both NVM and yarn was necessary. Making this refresh explicit makes it so we can re-use this command on both Mac and Linux. 

Alternatively, we could revert this to the original and I can create an isolated node module installation step for Mac only.

nsakaimbo committed:
remove unnecessary MINIFY_JS env var

nsakaimbo commented on .circleci/config.yml:
```diff
@@ -14,6 +14,15 @@ references:
+      - image: circleci/node:12.18.0-browsers
+    environment:
+      CONFIG_ENV: release
+      MINIFY_JS: true
```
Removed in 88695e6.

nsakaimbo committed:
cache only contents of desktop folder

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure (ci/wp-desktop): @nsakaimbo please re-try this workflow ("Rerun Workflow from Failed") and/or review this PR for breaking changes. Please also ensure this branch is rebased off latest Calypso master.

bsessions85 reviewed (approved):
Looks good to me. nice work!

jsnajdr commented on .circleci/config.yml:
```diff
@@ -574,6 +615,74 @@ jobs:
+            rm -rf desktop/release/mac
+      - store_artifacts:
+          when: always
+          path: desktop/release
```
Good to see that we're storing the built app package as an artifact :+1:

jsnajdr commented on package.json:
```diff
@@ -62,12 +62,14 @@
-		"build-desktop:install-app-deps": "check-npm-client && cd desktop && yarn install --frozen-lockfile",
-		"build-desktop:app": "check-npm-client && cd desktop && yarn run build",
+		"build-desktop:install-app-deps": "cd desktop && yarn install --frozen-lockfile",
+		"build-desktop:app": "cd desktop && yarn run build",
```
Did the `check-npm-client` cause any mischief here? We probably don't need to run the check all the time, especially in sub-tasks that are not meant to run directly from the CLI, but I'm just curious why you removed it.

nsakaimbo replied:
> Did the check-npm-client cause any mischief here?

The `build-desktop:app` script needs to be executed within a platform-specific (Windows/Mac/Linux) context where Calypso's modules are not installed. `check-npm-client` is not one of the desktop's dependencies, and so this script would fail. I think this is fine, however - the `build-desktop:source` script is prefixed with the `check-npm-client`, so we know this check has occurred upstream by the time the Electron packaging step occurs.

jsnajdr commented on package.json:
```diff
@@ -62,12 +62,14 @@
 		"build-css:reader-mobile": "check-npm-client && node-sass --importer node_modules/node-sass-package-importer/dist/cli.js --include-path client client/assets/stylesheets/reader-mobile.scss public/reader-mobile.css --output-style compressed && yarn run -s postcss public/reader-mobile.css",
 		"build-devdocs:search-index": "check-npm-client && node bin/generate-devdocs-search-index.js",
-		"build-desktop": "check-npm-client && run-s build-desktop:config build-desktop:server build-desktop:client build-desktop:install-app-deps build-desktop:app",
+		"build-desktop": "check-npm-client && yarn run build-desktop:source && run-s build-desktop:install-app-deps build-desktop:app",
```
Can we run all three `build-desktop:*` tasks inside a single `run-s`? It effectively does the same thing as `yarn run a && yarn run b && yarn run c`, waiting for one job to finish before starting the next one.

nsakaimbo replied:
Updated in 57921dd56cfb7705d9f51af83ff774bbc25fea38.

jsnajdr reviewed (approved):
This looks good 🙂

I still suspect that the `nvm` and `yarn` setup could be greatly simplified, but that's a research task for a followup optimization PR.

Thanks for working on the second half of the `wp-desktop` import!

nsakaimbo committed:
use run-s in build-desktop script

stephanethomas commented:
Hello @nsakaimbo 👋 . It looks like some of the files introduced in this pull request were made executable:

```shell
desktop/e2e/tests/lib/

drwxrwxr-x 2 vagrant vagrant 4096 2020-07-10 10:07 components
drwxrwxr-x 2 vagrant vagrant 4096 2020-07-10 10:07 pages
-rwxrwxr-x 1 vagrant vagrant 1093 2020-07-10 10:07 async-base-container.js
-rw-rw-r-- 1 vagrant vagrant 1026 2020-07-10 10:07 data-helper.js
-rwxrwxr-x 1 vagrant vagrant 7786 2020-07-10 10:07 driver-helper.js
-rw-rw-r-- 1 vagrant vagrant  222 2020-07-10 10:07 mocha-hooks.js
-rw-rw-r-- 1 vagrant vagrant 1459 2020-07-10 10:07 video-recorder.js

desktop/e2e/tests/lib/components/

-rw-rw-r-- 1 vagrant vagrant  790 2020-07-10 10:07 editor-confirmation-sidebar-component.js
-rwxrwxr-x 1 vagrant vagrant  813 2020-07-10 10:07 nav-bar-component.js
-rw-rw-r-- 1 vagrant vagrant 2338 2020-07-10 10:07 post-editor-toolbar-component.js
-rwxrwxr-x 1 vagrant vagrant 1516 2020-07-10 10:07 post-preview-component.js
-rw-rw-r-- 1 vagrant vagrant  572 2020-07-10 10:07 sidebar-component.js

desktop/e2e/tests/lib/pages/

-rw-rw-r-- 1 vagrant vagrant  539 2020-07-10 10:07 checklist-page.js
-rw-rw-r-- 1 vagrant vagrant 1431 2020-07-10 10:07 checkout-page.js
-rw-rw-r-- 1 vagrant vagrant 1397 2020-07-10 10:07 editor-page.js
-rwxrwxr-x 1 vagrant vagrant 1717 2020-07-10 10:07 login-page.js
-rwxrwxr-x 1 vagrant vagrant  667 2020-07-10 10:07 profile-page.js
-rwxrwxr-x 1 vagrant vagrant  330 2020-07-10 10:07 reader-page.js
-rw-rw-r-- 1 vagrant vagrant 3261 2020-07-10 10:07 signup-steps-page.js
-rwxrwxr-x 1 vagrant vagrant 1112 2020-07-10 10:07 view-post-page.js
```

As far I can see this is not needed, could you fix this when you get the chance? Thanks!

nsakaimbo commented:
Hi, @stephanethomas - Thanks for the heads up! That's really, _really_ odd - not sure how that would happen. Maybe some permissions weirdness when I was copying files over locally. Will fix in a follow-up PR.

stephanethomas commented:
Hello hello @nsakaimbo 👋. Did you managed to fix those permissions? I've just checked out a fresh copy of the repository, and those files still have the wrong flag.

nsakaimbo commented:
👋 @stephanethomas Apologies this took me so long - thanks for checking in. I amended the file permissions in #47021 - please let me know if I missed anything.
